lib/github_advisory_sync.rb

```diff
@@ -0,0 +1,281 @@
+
+    def initialize(github_advisory_graphql_object:)
+      @github_advisory_graphql_object = github_advisory_graphql_object
+    end
+
+    # extract the CVE identifier from the GitHub Advisory identifier list
+    def cve_id
+      identifier_list = github_advisory_graphql_object["identifiers"]
+      cve_id_obj = identifier_list.find { |id| id["type"] == "CVE" }
+      return nil unless cve_id_obj
+
+      cve_id_obj["value"]
+    end
+
+    # return a date as a string like 2019-03-21.
+    def published_day
+      return nil unless github_advisory_graphql_object["publishedAt"]
+
+      pub_date = Date.parse(github_advisory_graphql_object["publishedAt"])
+      # pub_date.strftime("%Y-%m-%d")
+      pub_date
+    end
+
+    def package_names
+      github_advisory_graphql_object["vulnerabilities"]["nodes"].map{|v| v["package"]["name"]}.uniq
+    end
+
+    def rubysec_filenames
+      package_names.map do |package_name|
+        File.join("gems", package_name, "#{cve_id}.yml")
+      end
+    end
+
+    def withdrawn?
+      !github_advisory_graphql_object["withdrawnAt"].nil?
+    end
+
+    def external_reference
+      github_advisory_graphql_object["references"].first["url"]
+    end
+
+    def vulnerabilities
+      github_advisory_graphql_object["vulnerabilities"]["nodes"]
+    end
+
+    def has_ruby_vulnerabilities?
+      vulnerabilities.any? do |vuln|
+        vuln["package"]["ecosystem"] == "RUBYGEMS"
+      end
+    end
+
+    def some_rubysec_files_do_not_exist?
+      rubysec_filenames.any?{|filename| !File.exist?(filename) }
+    end
+
+    def write_files
+      return [] unless cve_id
+      return [] unless some_rubysec_files_do_not_exist?
+
+      files_written = []
+      vulnerabilities.each do |vulnerability|
+        filename_to_write = File.join("gems", vulnerability["package"]["name"], "#{cve_id}.yml")
+        next if File.exist?(filename_to_write)
+
+        data = {
+          "gem" => vulnerability["package"]["name"],
+          "cve" => cve_id[4..20],
+          "date" => published_day,
+          "url" => external_reference,
+          "title" => github_advisory_graphql_object["summary"],
+          "description" => github_advisory_graphql_object["description"],
+          "cvss_v3" => "<FILL IN IF AVAILABLE>",
+          "patched_versions" => [ "<FILL IN SEE BELOW>" ],
+          "unaffected_versions" => [ "<OPTIONAL: FILL IN SEE BELOW>" ]
+        }
+
+        dir_to_write = File.dirname(filename_to_write)
+        Dir.mkdir dir_to_write unless Dir.exist?(dir_to_write)
+        File.open(filename_to_write, "w") do |file|
+          # create an automatically generated advisory yaml file
+          file.write data.to_yaml
+
+          # The data we just wrote is incomplete,
+          # and therefore should not be committed as is
+          # We can not directly translate from GitHub to rubysec advisory format
+          #
+          # The patched_versions field is not exactly available.
```

Comment:
Can you provide some examples of what you see here? I wonder if we could guess a bit, knowing that it still needs human review but might make the process easier?

Reply:
so it is definitely possible to make some logic to automate the most common cases of translation.

Some common cases:
```
      {
          "package": {
            "name": "chloride"
          },
          "vulnerableVersionRange": "< 0.3.0",
          "firstPatchedVersion": {
            "identifier": "0.3.0"
          }
        },
```

```
        {
          "package": {
            "name": "actionview"
          },
          "vulnerableVersionRange": ">= 4.0.0, < 4.2.11.1",
          "firstPatchedVersion": {
            "identifier": "4.2.11.1"
          }
        },
        {
          "package": {
            "name": "actionview"
          },
          "vulnerableVersionRange": ">= 5.2.0, < 5.2.2.1",
          "firstPatchedVersion": {
            "identifier": "5.2.2.1"
          }
        },
```

```
        {
          "package": {
            "name": "actionview"
          },
          "vulnerableVersionRange": ">= 4.0.0, < 4.2.11.1",
          "firstPatchedVersion": {
            "identifier": "4.2.11.1"
          }
        },
        {
          "package": {
            "name": "actionview"
          },
          "vulnerableVersionRange": ">= 5.2.0, < 5.2.2.1",
          "firstPatchedVersion": {
            "identifier": "5.2.2.1"
          }
        },
```

I think my preference would be to follow this logic up with another PR.  We might need to add some tests for the smart logic like that and this isn't really structured right now for that